io/src/main/scala/fs2/io/tcp/Socket.scala

```diff
@@ -177,90 +177,128 @@ protected[tcp] object Socket {
+      // of (bytes_read, read_duration)
+      def readChunk(buff:ByteBuffer, timeoutMs:Long):F[(Int,Long)] = F.async { cb => F.pure {
+        val started = System.currentTimeMillis()
+        ch.read(buff, timeoutMs, TimeUnit.MILLISECONDS, (), new CompletionHandler[Integer, Unit] {
+          def completed(result: Integer, attachment: Unit): Unit =  {
+            val took = System.currentTimeMillis() - started
+            F.unsafeRunAsync(F.start(F.delay(cb(Right((result, took))))))(_ => ())
+          }
+          def failed(err: Throwable, attachment: Unit): Unit = F.unsafeRunAsync(F.start(F.delay(cb(Left(err)))))(_ => ())
+        })
+      }}
+
+      // gets buffer of desired capacity, ready for the first read operation
+      // If the buffer does not have desired capacity it is resized (recreated)
+      // buffer is also reset to be ready to be written into.
+      def getBufferOf(sz: Int): F[ByteBuffer] = {
+        bufferRef.get.flatMap { buff =>
+          if (buff.capacity() < sz) bufferRef.modify { _ => ByteBuffer.allocate(sz) } map { _.now }
+          else {
+            buff.clear()
+            F.pure(buff)
+          }
         }
-        def failed(err: Throwable, attachment: Unit): Unit = F.unsafeRunAsync(F.start(F.delay(cb(Left(err)))))(_ => ())
-      })
-    }}
+      }
 
-    def read0(max:Int, timeout:Option[FiniteDuration]):F[Option[Chunk[Byte]]] = {
-      val buff = ByteBuffer.allocate(max)
-      readChunk(buff,timeout.map(_.toMillis).getOrElse(0l)).map {
-        case (read,_) =>
-          if (read < 0) None
-          else Some(Chunk.bytes(buff.array(), 0, read))
+      // When the read operation is done, this will read up to buffer's position bytes from the buffer
+      // this expects the buffer's position to be at bytes read + 1
+      // Additionally this releases the semaphore to allow for other read operation to take a place
+      def releaseBuffer(buff: ByteBuffer): F[Chunk[Byte]] = {
+        val read = buff.position()
+        if (read == 0) readSemaphore.increment >> F.pure(Chunk.bytes(Array.empty))
+        else {
+          val dest = Array.ofDim[Byte](read)
+          buff.flip()
+          buff.get(dest)
+          readSemaphore.increment >> F.pure(Chunk.bytes(dest))
+        }
       }
-    }
 
-    def readN0(max:Int, timeout:Option[FiniteDuration]):F[Option[Chunk[Byte]]] = {
-      def go(buff: ByteBuffer, timeoutMs:Long):F[Option[Chunk[Byte]]] = {
-        readChunk(buff,timeoutMs).flatMap { case (readBytes, took) =>
-          if (readBytes < 0 || buff.remaining() <= 0) F.pure(Some(Chunk.bytes(buff.array(), 0, buff.position())))
-          else go(buff,(timeoutMs - took) max 0)
+      def read0(max:Int, timeout:Option[FiniteDuration]):F[Option[Chunk[Byte]]] = {
+        readSemaphore.decrement >>
+        getBufferOf(max) flatMap { buff =>
+          readChunk(buff, timeout.map(_.toMillis).getOrElse(0l)) flatMap {
+            case (read, _) =>
+              if (read < 0) F.pure(None)
```

Comment:
but actually, hence this signals eof, i think we won't hit it,neverthelss I gonna fix it for corectness